### PR TITLE
[Impeller] avoid locking fence waiter while querying fence status.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Checking the fence status has been recorded to take up to 70ms on the S10 devices. Change the behavior of fence waiter so that it pulls all fences/cbs into temporaries, then movies any unresolved fences back after the wait.

Fixes https://github.com/flutter/flutter/issues/129371

### CODE REVIEW

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc [hunk 1]
```diff
@@ -54,56 +53,36 @@ void FenceWaiterVK::Main() {
...
     }
 
     auto result = device_holder->GetDevice().waitForFences(
-        wait_set->size(),                        // fences count
-        wait_set->data(),                        // fences
-        false,                                   // wait for all
+        fences.size(),                           // fences count
+        fences.data(),                           // fences
+        true,                                    // wait for all
         std::chrono::nanoseconds{100ms}.count()  // timeout (ns)
```
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:77]: Should we adjust the timeout now that we are waiting on all of them?  100ms / fence?

**Review by gaaclarke — COMMENTED** [review]
logic lgtm, it's hard to guess what downstream effects this could have since shifting the deletion of resources to be later can have knock on memory problems in these limited memory devices.  chinmay probably has a better idea about that.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:77]: i'd defer to @chinmaygarde on that, I left the old default but all the values seem pretty arbitrary. I guess if we're waiting for all of them,it makes sense to make it longer but not sure how long.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc [hunk 2]
```diff
@@ -54,56 +53,36 @@ void FenceWaiterVK::Main() {
...
     );
     if (!(result == vk::Result::eSuccess || result == vk::Result::eTimeout)) {
```
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:79]: Previously if this was an error, the fence would be tried again later right?  Now it gets lost.

Comment (gaaclarke) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:79]: nevermind, it would get lost in the old code too.

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:79]: it doesn't get lost, everything is dropped at the end of the scope here. The only other documented error condition is device loss which makes all of these invalid anyway.

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:79]: so barring the issue with timing out, I don't think there is a problem.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:77]: maybe if we hit the timeout condition, I grab the mutex again and add all the fences back into the set?

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
I think we should just move the checks and destructors out of the critical section instead. That way, we are eagerly collecting resources as much as possible. I am prototyping this now.

**Comment by jonahwilliams** [comment]
Happy to take another approach here, but

>  That way, we are eagerly collecting resources as much as possible.

We're eventually not going to be doing that, because everything I've read on performance guidlines says we _must_ minimize the number of queue submits, which means we'll eventually get to a point where we have ~one fence per frame.

Of course with this design we may end up waiting on fences from multiple frames, which is probably no bueno.

**Comment by gaaclarke** [comment]
> which means we'll eventually get to a point where we have ~one fence per frame.

That's what I hacked together here: https://github.com/flutter/engine/commit/ffdccce06edbd8907743ca737a524f85680862ea

I was about to write up a design doc suggesting this but when I measured that I didn't see a big difference in performance.  What I found is there are queue submits that are happening that are outside the scope of drawing a frame too, so I only removed the fences related to drawing the frame.

Based on what you are seeing with respect to waiting on the fences it seems like that proposal might actually help by eliminating fences.  Since you can reproduce the problem could you try out my branch?

**Comment by gaaclarke** [comment]
I chatted with @jonahwilliams offline.  I wasn't realizing that the other PR he has that is reducing vkQueueSubmit calls is also getting rid of fences and that sounds like the good path to follow (not just reducing fences but also reducing vkQueueSubmit calls).

We also discussed other the queue submits for things like blitting or mippmapping that exist outside of the frame and we didn't have a good suggestion yet for those.  I had accidentally batched those in my commit, but it will take some thought to get them cleaning up correctly if we wanted to tie them to a frame.

**Comment by jonahwilliams** [comment]
Okay, I've pushed up a version that doesn't change _how_ we wait, but holds the lock for a shorter duration by moving all of the fences into a temporary and then by pushing the unsignaled fences back into the wait set afterwards

**Comment by chinmaygarde** [comment]
I noticed a couple of minor issues that I'm not sure matter but were bothersome anyway.
* The callbacks were invoked within the critical section. This is problematic as an API user because you can't add more fences to the waiter from the callback. And, if/when the callbacks hit allocators, they may be slow.
* Waiter collection doesn't invoke the callback so there may be some potential leaks in shutdown.

I attempted to fix both in https://github.com/flutter/engine/pull/43131.

**Comment by jonahwilliams** [comment]
Closing in favor of that one 😄

**Pull request closed without merging**